nucleus/modelci/eval_functions/available_eval_functions.py

```diff
@@ -0,0 +1,81 @@
+from typing import List, Dict, Union, Type
+
+from nucleus.modelci.eval_functions.base_eval_function import BaseEvalFunction
+from nucleus.modelci import EvalFunctionEntry
+
+MEAN_AVG_PRECISION_NAME = "mean_average_precision_boxes"
+IOU_NAME = "IOU"
+
+
+class IoU(BaseEvalFunction):
+    """Intersection over union for all bounding boxes"""
+
+    @classmethod
+    @property
+    def name(self) -> str:
+        return "IOU"
+
+
+class MeanAveragePrecisionForBoxes(BaseEvalFunction):
+    """Mean average precision for boxes fitting the provided parameters"""
+
+    @classmethod
+    @property
+    def name(self) -> str:
+        return "mean_average_precision_for_boxes"
+
+
+EvalFunction = Union[Type[IoU], Type[MeanAveragePrecisionForBoxes]]
+
+
+class AvailableEvalFunctions:
+    """Collection class that acts as a common entrypoint to access evaluation functions. Standard evaluation functions
+    provided by Scale are attributes of this class.
+
+    Examples:
+        e = client.modelci.eval_functions
+        unit_test_metrics = [e.iou() > 5, e.map() > 0.95, e.custom["customer_function"]() == True]
+    """
```

Comment:
I get the idea here, but I think it's a bit confusing and unscalable to have a separate class that acts as a collection of all the eval functions. I'd prefer to keep a native list for consistency since all of our other "list" methods return native lists.

Reply:
I think in this case, the functionality to map by name is sort of optional for users, and we can provide optional functionality in the form of util functions.

Reply:
A list or a dictionary doesn't play well at all with an IDE or Jupyter Notebook. You can't return a named entity without having an aggregator class. 

In this instance the `eval_functions` property behaves more like a namespaced module than a list ... and I think that's good and Pythonic.

When we add new evaluation functions we should add them as named objects here as well to make sure that the Jupyter notebook/IDE knows what available eval functions there are with normal autocomplete functionality. The abstraction `AvailableEvalFunctions` should barely be seen by the user, they only interact with its attributes through the modelci class.

I actually would like to change our other list methods into returning a query object as well, similar to this one. That allows for a lot more flexibility in interacting with the backend in a fast manner. 

I think this interface is very user-friendly and easier to write than more REST like APIs so I suggest we try this out and if we get bad feedback we can roll back to a simpler model.

Reply:
I think it's a bit annoying that we need to add info about new eval functions at many different places (new class, add to `EvalFunctions`, add to `AvailableEvalFunctions`). But I guess whenever we try not to have the eval functions as member of `AvailableEvalFunctions`, IDEs won't like it, right?